feat(windows): Allow perma-caching stable path dirs to achieve low input lag

eugenesvk opened the pull request:
This PR (with others) finally allows to have no input lag on Windows!!!

(It's on top of https://github.com/xonsh/xonsh/pull/5654 for the same reasons)

It allows caching executables in **large non-changing dirs** like `C:\Windows\System32` (user configurable) to avoid any IO during per-keystroke checks for color highlighting
(though the actual cost doesn't really depend on the size of the dir, it's the same 15+ `"file.ext".is_file()` for every `pathext`)

- Also allows caching (by mtime) current working dir
  - since (to avoid per-file IO) non-executables are also cached, pyghooks with highlight prefixes for any file, not only commands
  - this can be user-fixed by caching executables, but then it will negate the benefit of no per-file IO (at least once per every newly seen dir per session)

As a bonus it adds color highlighting for partial commands (for commands in this perma cache, session cache, and dirs listed every time) via a new `Token.Name.Cmdprefix` token scope (default is white underlines), which is enabled by the new dependency [pygtrie](https://pypi.org/project/pygtrie/) that creates a character-by-character visibility into the cache list, so for every typed string you know whether there is a command that starts with that string (as discussed in https://github.com/xonsh/xonsh/issues/5541), this could also be expanded into other command cache types
And since this cool syntax check is helpful on other platforms as well, I've removed the "Windows" limits from the caches


https://github.com/user-attachments/assets/b8138c98-614a-4b94-9696-e5e80f4b0a15



- [x] added news
- [x] doc: added to the platform issues
- [x] example: cache dirs just imrov perf
- [x] Addressing https://github.com/xonsh/xonsh/issues/5612 and also some ideas from https://github.com/xonsh/xonsh/issues/5541 and a suggested feature from https://github.com/xonsh/xonsh/discussions/5653
- [x] added cache and `Cmdprefix` color highlighting

**Questions** 
1. currently there are still 2 IO checks - if a name if found in permacache, it still checks whether a file exists (though twice)
Wouldn't it be ok to just assume it does since it's `perma` cache, so assumed to not change?
Added a new config var to allow skipping that

2. How to set the style for the new partial highlighting to only be underline without specifying the color? I've tried just setting `underline`, but then the color changes. And if I specify `underline ansiwhite` as currently, then I lose the light-gray color on hyphens
 
3. Is there an example from xonsh code I could copy to make permacache generation a non-blocking operation as it's potentially expensive to list System32 folders (though it should only happen once, so maybe not a big deal)

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

eugenesvk committed:
add news

eugenesvk committed:
add news

eugenesvk committed:
add 'XONSH_WIN_PATH_DIRS_TO_LIST' config env variable

eugenesvk committed:
test: add test_xonsh_win_path_dirs_to_list

eugenesvk committed:
List dirs in PATHs for faster typing (color highlighting)

If user defined XONSH_WIN_PATH_DIRS_TO_LIST, then list the files in those dirs directly instead of checking for the existence of every single possible executable (10+ checks) directly, which for smaller dirs of a few dozen files (which is most of dirs in PATH) is faster

eugenesvk committed:
doc: add a Windows tip

pre-commit-ci[bot] committed:
test: add more iterations to avoid CI failing sometimes

reformat

eugenesvk committed:
consolidate dupe code into a separate function

eugenesvk committed:
add an argument to allow disabling use of path cache to locate_executable

eugenesvk committed:
pre-commit

eugenesvk committed:
add an argument to enable using dir session cache

eugenesvk committed:
use session dir cache for syntax highlighting if configured

eugenesvk committed:
simplify executable list of conditional branches

eugenesvk committed:
rename PathCleanCache to PathClean to also store dir session cache

eugenesvk committed:
add dir_cache to PathCache

eugenesvk committed:
add news

eugenesvk committed:
pre-commit fixes

eugenesvk committed:
doc: update platform tips

eugenesvk committed:
add test for xonsh_win_dir_session_cache

eugenesvk committed:
fix condition check errors

eugenesvk committed:
add type info

eugenesvk committed:
add lower-case set for matching file existence in listed dirs

eugenesvk committed:
exe search: compare lowercased file names with lowercased cache/dir list

eugenesvk committed:
Return case sensitive found file output when we have such information:

Use two lists when scanning a dir (or using a cache of a dir scan):
- case insensitive for matching a file in a list of files
- case sensitive for output

eugenesvk committed:
update function comment

eugenesvk committed:
pre-commit fixes

eugenesvk committed:
skip a timing test when for some reason 0 duration is reported

eugenesvk committed:
cache the PATH used to cache clean_paths

Avoids returning the wrong cache for a locate file call with env.swap(PATH=['a'])

eugenesvk committed:
refactor clean path cache to store multiple clean paths

eugenesvk committed:
add a function to hash a list of strings

eugenesvk committed:
pre-commit

eugenesvk committed:
refactor timing in test_locate_file_clean_path_cache_time to avoid CI issues

where timing is sometimes just 0

eugenesvk committed:
don't expand symlinks when cleaning paths

avoids breaking cache when symlinked $PATH is set by the user to be cached, but after being resolved to a non-symlinked path it no longer matches what the user wants

eugenesvk committed:
doc: add a note on normalizing path to match internally cleared paths for cache

eugenesvk committed:
update function comment

eugenesvk committed:
style: underline valid command prefixes

eugenesvk committed:
add support for partial command highlighting

eugenesvk committed:
move executables-related commands to executables from commands cache

since they're used in paths cache as well and paths cache can't import commands_cache due to circular imports

eugenesvk committed:
convert PathCache to a singleton

eugenesvk committed:
PathCache: add support for perma cached dirs stored in a file and valid across sessions

eugenesvk committed:
do case insensitive partial match

eugenesvk committed:
add use_perma_cache arg

eugenesvk committed:
add partial_match argument to pyghooks/path searches to signal if typed text is a prefix of a valid command

currently only works for the few perma cached dirs, but can be expanded in the future to any command list that implements a trie

eugenesvk committed:
update comment

eugenesvk committed:
add news

eugenesvk committed:
disable a stub for creating a cache of all commands

currently only a few perma-cached dirs are supported

eugenesvk committed:
add test

eugenesvk committed:
pre-commit test

eugenesvk committed:
pre-commit

eugenesvk committed:
ruff

eugenesvk committed:
remove leftover dbg

eugenesvk committed:
dep: add pygtrie

eugenesvk committed:
fix missing partial_match arg and a type check

eugenesvk committed:
doc: fix a typo

eugenesvk committed:
doc: add to platform issues

eugenesvk committed:
add type annotation

eugenesvk committed:
fix type mismatch

eugenesvk committed:
use missing use_perma_cache arg

eugenesvk committed:
use use_dir_session_cache arg

eugenesvk committed:
test: use use_dir_session_cache arg

eugenesvk committed:
fix path perma cache being reset on each load instead of being initialized only once

eugenesvk committed:
add type annotation

eugenesvk committed:
allow skipping the is_file check

helpful when we know that a file exists, e.g., we've got the file from listing a dir or we matched in a dir listing cache and are ok with cache being the "source of truth"

eugenesvk committed:
add XONSH_WIN_DIR_CACHE_SKIP_EXIST to allow skipping a is_file check in cached dirs

useful for faster syntax highlighting

eugenesvk committed:
don't check that a file exists for dirs we scan every time

since we already get a list of existing files

eugenesvk committed:
reformat

eugenesvk committed:
fix argument position

eugenesvk committed:
doc: add description for XONSH_WIN_DIR_CACHE_SKIP_EXIST

eugenesvk committed:
remove unused condition

eugenesvk committed:
add support for signaling partial matches to temporarily cache dirs

from XONSH_WIN_PATH_DIRS_TO_LIST and XONSH_WIN_DIR_SESSION_CACHE

eugenesvk committed:
test: add a test for pyghooks Cmdprefix parsing

eugenesvk committed:
replace partial_match list with a class

eugenesvk committed:
add missing dir perma cache argument to pygments

eugenesvk commented:
And I've already addressed that note

eugenesvk committed:
add PathCache comment

eugenesvk committed:
add a cache for smaller dirs listed on each keystroke

eugenesvk committed:
update comment

eugenesvk committed:
doc: remove the need to normalize by the user

eugenesvk committed:
reformat

eugenesvk committed:
style valid prefixes only while they're still prefixes, not when it's a standalone word

- 'xon' should be styled as a valid prefix to aid further typing
- 'xon = 2' shouldn't be

eugenesvk committed:
add support for dir perma cache to locate file

with a signal of partial match

eugenesvk committed:
do not scan perma cache dirs again if nothing was found there

eugenesvk committed:
doc: fix a typo

eugenesvk committed:
rename XONSH_WIN_DIR_SESSION_CACHE to XONSH_DIR_SESSION_CACHE

eugenesvk committed:
doc: update XONSH_DIR_SESSION_CACHE comment with a highlighting example

eugenesvk committed:
don't limit to Windows

since this can be used for future typo syntax highlighting on other platforms

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Spitfire1900 commented:
The best test for this is to connect to a Netshare folder over a slow network. This is definitely faster yet is still has opportunities for improvement.

In the optimal case there should be no blocking item before accepting a keypress, all highlighting cases are run on a different thread and update the highlighting when available.

eugenesvk committed:
test: add timing test for PathCleanCache

eugenesvk committed:
cache a list of cleaned paths to avoid wasting time cleaning it on every keystroke

but only for the default XSH environment

cache is reset on new prompt (before every on_pre_prompt_format is fired)

eugenesvk commented:
> This is definitely faster yet is still has opportunities for improvement.

Have you tried adding all PATHs to e.g. a session cache? Then it should have no IO per keystroke for highlighting; or maybe it bugs somewhere, or maybe xonsh does some other checks? Think the local folder check that can add lag only happens if you type `./`, but don't remember

> In the optimal case there should be no blocking item before accepting a keypress, all highlighting cases are run on a different thread and update the highlighting when available.

Sure, these PRs aren't that, they can only achive no lag at some loss of precision. There is an open issue from a few years back where the solution was to do all this highlighting stuff in its rightful place - off the main thread...

eugenesvk committed:
add XONSH_WIN_DIR_SESSION_CACHE config env var to signal a list of dirs to cache a scan of

eugenesvk committed:
use dir_cache if configured

eugenesvk committed:
update locate_file_in_path_env description

eugenesvk committed:
don't limit to Windows only

since future-implemented syntax highlighting is useful everywhere

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

eugenesvk committed:
add XONSH_WIN_DIR_PERMA_CACHE env var

eugenesvk committed:
doc: fix a typo

eugenesvk committed:
use more consistent names

- use_dir_cache_perma  , dir_cache_perma
- use_dir_cache_session, dir_cache_session

eugenesvk committed:
clean up user config vars with lists of PATHs on load

eugenesvk committed:
don't limit to Windows only

since prefix syntax highlighting is useful everywhere

eugenesvk committed:
doc: update env var description

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

eugenesvk committed:
de-win dir perma cache file

Spitfire1900 commented:
> > This is definitely faster yet is still has opportunities for improvement.
> 
> Have you tried adding all PATHs to e.g. a session cache? Then it should have no IO per keystroke for highlighting; or maybe it bugs somewhere, or maybe xonsh does some other checks? Think the local folder check that can add lag only happens if you type `./`, but don't remember

Are these comma separated?
https://github.com/xonsh/xonsh/pull/5660/files#diff-545bb312bf20b24ad2654671db11758ff3504266bd7521e0b0ce43558603fac3

eugenesvk commented:
Those are regular python sets/lists, so each dir is a separate element

eugenesvk committed:
test: fix env var typo

eugenesvk committed:
fix env var typo

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Spitfire1900 commented:
@eugenesvk is the following valid configuration in `.xonshrc`?

```xonsh
$XONSH_DIR_SESSION_CACHE = $PATH[-1] # add everything but current directory to Session cache.
$XONSH_DIR_CACHE_SKIP_EXIST = True
```

eugenesvk commented:
But your current directory isn't part of $PATH, so -1 doesn't make sense.
Anyway, those are invalid, you'd need a list of actual dirs, these lists are static as they were meant to be used to handle static PATH var 

Interesting idea about changing the  logic for handling of the current dir, but note that it's not guaranteed to be faster, e.g. for large dirs listing would be slower than checking for 15.pathext. Though maybe if you cache it might still be worth it, especially if you don't expect to visit very large dirs often?

eugenesvk committed:
typo

eugenesvk committed:
doc: add missing spaces to env vars

eugenesvk committed:
add XONSH_DIR_CACHE_LIST_EXT_MIN to allow forcing a dir list on non-Windows

and benefit from partial prefix command syntax highlighting

eugenesvk committed:
test: enable cmdprefix outside of Windows

eugenesvk committed:
ci: use current 3.13 version of python (#5751)

pre-commit: use current version of python

eugenesvk committed:
fix: excessive permanent cache writing due to an equality fail

"pathext == self._pathext_cache" fails for empty []≠set() while it should "logically" succeed,
so add an extra check that pathext isn't empty

eugenesvk committed:
add a reset method to PathCache singleton

some tests reset env, so a previously created singleton with an old env will not be reinitialized when env is restored, thus leading to tests failing
This would allow tests to reset it before running

eugenesvk committed:
tests: fix failing due to PathCache being initialized with a different env in other tests

Spitfire1900 commented:
See https://github.com/xonsh/xonsh/pull/5677#issuecomment-2538930507's attached video  for performance differences in a worst case scenario (corporate on-access antivirus scanner)

https://github.com/user-attachments/assets/7ca0e18f-17cf-46b2-a0b5-74e0deb14f27

Spitfire1900 commented:
>But your current directory isn't part of $PATH, so -1 doesn't make sense (but should be valid, it should just expand to whatever your startup $PATH is less its last element and added as a list of dirs to cache).

Any empty string is the last entry in `$PATH`
```xonsh
@ $PATH[-1]
''
@ 
''`

Spitfire1900 commented:
@eugenesvk is there a good way to check the state of the cache while in an interactive xonsh session?

eugenesvk commented:
well, yeah, you're not caching anything from your quoted config, all the cache vars are empty

`$PATH[-1]` maybe that's just a Windows issue with parsing the default path that it leaves the last one empty, but just print the full one`$PATH`, copy all paths from there and add then to one of the 3 cache dirs, you can temporarily add all of them to the permacache just for testing. Checking current dir isn't related to your $PATH var

Re. speed comparison: the best one would be
- do 2 videos of just a single prompt line with recording started one you press & hold 1 key (no need to show OSD, it doesn't help gauge the lag)
- put these videos one on top of the other 
- while recording, press play at the same time (e.g., one with a key, another with a mouse)
- then you'll se by the distance the spammed key travels how much faster one is

Similar to this: https://github.com/xonsh/xonsh/issues/5612 (though this adds 2 videos separately, but if you click play at the same time you'll see one being faster)

>  is there a good way to check the state of the cache while in an interactive xonsh session?

Nothing comprehensive, there is a singleton `PathCache` which you could inspect, or use some rudimentary summary `from xonsh.procs.executables import PathCache; pc = PathCache(None); pc.get_cache_info()` (though this one hasn't been pushed here yet)

and compare to `$XONSH_DIR_PERMA_CACHE; $XONSH_DIR_SESSION_CACHE; $XONSH_DIR_CACHE_TO_LIST` to see which dirs you wanted cache

eugenesvk committed:
typo

eugenesvk committed:
add get_cache_info for runtime debugging

eugenesvk committed:
cache last path hash to avoid IO for clearing paths when path hash changes.

PathCache.is_dirty is reponsible for updating the clear paths cache and path hash (on each prompt)

eugenesvk committed:
fmt

eugenesvk committed:
use the new get_clean_path in cache update

eugenesvk committed:
doc: add info on get_cache_info

eugenesvk committed:
doc: fix for old pythons

eugenesvk committed:
test: fix wrong type

eugenesvk committed:
include user dirs not in PATH on startup

they can appear in PATH later (even some plugins can add on startup too late for thie code to catch it), so no point in not checking them later

eugenesvk committed:
doc: reformat

eugenesvk committed:
add path_cache_dirty to avoid breaking tests

and allow pyghooks to explicitly opt into using a cache of Env's PATH even if it chanes mid-line (new prompts reset cache anyway)

eugenesvk committed:
test: add pathext closer to default for more reliable performance comparison

eugenesvk committed:
doc: add dir cache vars for current dirs

eugenesvk committed:
doc: update get_cache_info with CWD cache

eugenesvk committed:
feat: add support for caching current working dir

eugenesvk committed:
fmt

Spitfire1900 commented:
What triggers whether un-cached gets cached?

```
@ PathCache(xonsh.environ).get_cache_info()
PATH    : ∑  62 dirty
          └  54 clean (unique & existing)
Cached  : ∑   0 of which:              (pc = PathCache(None))
          ├   0 permanently            (pc.usr_dir_list_perma   ← $XONSH_DIR_PERMA_CACHE  )
          ├   0 this session           (pc.usr_dir_list_session ← $XONSH_DIR_SESSION_CACHE)
          └   0 by dir mtime, list onΔ (pc.usr_dir_list_key     ← $XONSH_DIR_CACHE_TO_LIST)
Uncached: ∑  54 including:
  C:\Program Files\Git\usr\bin
  C:\Program Files\Git\mingw64\bin
  C:\Program Files\Python313\Scripts
  C:\Program Files\Python313
  C:\Program Files\Eclipse Adoptium\jdk-17.0.13.11-hotspot\bin
  ...
```

eugenesvk committed:
add skip_exist support to cwd cache

eugenesvk commented:
> $XONSH_DIR_CACHE_TO_LIST = {''} # use the current directory as a list instead of checking for executablity.

You can now use `$XONSH_DIR_CWD_CACHE` and derivatives to add current working dirs to the cache using the same logic as XONSH_DIR_CACHE_TO_LIST, caching by mtime and listing on changes

eugenesvk commented:
> What triggers whether un-cached gets cached?

the 3 environment variables listed in the printed output. You need to actually add dirs you want to cache to them in your config

eugenesvk committed:
doc: update get_cache_info

eugenesvk committed:
fmt

eugenesvk committed:
fmt

eugenesvk committed:
add an option to cache only executables in mtime-cached dirs

eugenesvk committed:
doc: add XONSH_DIR_CACHE_LIST_NON_EXE

eugenesvk committed:
doc: update

eugenesvk committed:
fix wrong cache_non_exe is_exe checks

eugenesvk committed:
doc: update cache info message

eugenesvk committed:
move dir_cache_perma to a class var like other caches

eugenesvk committed:
add get_cache_db to get various db cache data

eugenesvk committed:
doc: update help info

eugenesvk committed:
fmt

eugenesvk committed:
feat: add cache file for listed dirs

XONSH_DIR_CACHE_TO_LIST + XONSH_DIR_CWD_CACHE

eugenesvk committed:
guard against empty session cache env vars

eugenesvk committed:
doc: update env var description

eugenesvk committed:
add missing imports

Spitfire1900 commented:
Yeah, running into issues here...

```xonsh
🐚 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh                                                                                                                                    cacheWinPathPerma   xonsh-cacheWinPathPerma  24-12-14 10:19:27-05
@ from xonsh.procs.executables import PathCache; pc = PathCache(None); pc.get_cache_info()
PATH    : ∑  66 dirty
          └  55 clean (unique & existing)
Cached  : ∑   1 of which:              (pc = PathCache(None))
          ├   1 permanently            (pc.usr_dir_list_perma   ← $XONSH_DIR_PERMA_CACHE       )
          ├   0 this session           (pc.usr_dir_list_session ← $XONSH_DIR_SESSION_CACHE     )
          └   0 by dir mtime, list onΔ (pc.usr_dir_list_key     ← $XONSH_DIR_CACHE_TO_LIST     )
Uncached: ∑  54 including:
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py", line 334, in get_cache_info
    +("✓" if env.get("XONSH_DIR_CWD_CACHE_NON_EXE", False) else "✗")
TypeError: bad operand type for unary +: 'str'
Return 1


🐚 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh                                                                                                                                    cacheWinPathPerma   xonsh-cacheWinPathPerma  24-12-14 10:19:45-05
@ $XONSH_DIR_PERMA_CACHE
['C:\\Program Files\\Git\\usr\\bin',
 'C:\\Program Files\\Git\\mingw64\\bin',
 'C:\\TMP\\xonsh-cacheWinPathPerma\\Scripts',
 'C:\\TMP\\xonsh-cacheWinPathPerma\\Scripts',
 'C:\\Program Files\\Git\\usr\\bin',
 'C:\\Program Files\\Git\\mingw64\\bin',
 'C:\\Program Files\\Python313\\Scripts\\',
 'C:\\Program Files\\Python313\\',
...
@ s = pickle.dumps(pathlib.Path('C:/Users/XXXXXXX/.cache/xonsh/dir_perma_cache.pickle').read_bytes())
@ import pprint

  obj = pickle.loads(s)
  pprint.pprint(obj)

(b'\x80\x04\x95`\x01\x00\x00\x00\x00\x00\x00]\x94(}\x94\x8c\x01.\x94\x8c\x07py'
 b'gtrie\x94\x8c\x08CharTrie\x94\x93\x94)\x81\x94}\x94(\x8c\x05_root'
 b'\x94h\x03\x8c\x05_Node\x94\x93\x94)\x81\x94]\x94(K\x0b\x8c\x01c\x94\x8c\x01o'
 b'\x94\x8c\x01n\x94\x8c\x01f\x94\x8c\x01t\x94\x8c\x01e\x94\x8c\x01s\x94h\x11h'
 b'\x02\x8c\x01p\x94\x8c\x01y\x94\x8c\x0bconftest.py\x94J\xf5\xff\xff\xff'
 b'K\x08h\x13h\x12h\x11\x8c\x01u\x94h\x14h\x02h\x14h\x15\x8c\x08setup.py\x94J'
 b'\xf8\xff\xff\xffK\x12\x8c\x01x\x94h\x0eh\x0fh\x13\x8c\x01h\x94\x8c\x01-\x94'
 b'\x8c\x01i\x94h\x0fh\x1b\x8c\x01d\x94h\x0eh\r\x8c\x01k\x94h\x12\x8c\x01'
 b'r\x94h\x02h\x14h\x15\x8c\x12xonsh-in-docker.py\x94eb\x8c\x07_sorted'
 b'\x94\x89ubs\x8f\x94(\x8c\x04.MSC\x94\x8c\x04.BAT\x94\x8c\x04.PYW'
 b'\x94\x8c\x04.CMD\x94\x8c\x04.WSF\x94\x8c\x04.VBS\x94\x8c\x04.VBE'
 b'\x94\x8c\x04.JSE\x94\x8c\x04.WSH\x94\x8c\x04.CPL\x94\x8c\x04.COM'
 b'\x94\x8c\x04.EXE\x94\x8c\x03.JS\x94\x8c\x03.PY\x94\x90e.')

```

`.xonshrc`:
```xonsh
# https://github.com/xonsh/xonsh/pull/5660#issuecomment-2537131523
$XONSH_DIR_PERMA_CACHE  = [i for i in $PATH]
$XONSH_DIR_CACHE_SKIP_EXIST  = True
$XONSH_DIR_CWD_CACHE = True
```

eugenesvk committed:
fix concat error

eugenesvk committed:
add XONSH_DIR_CACHE_LIST_SAVE_ON_EXIT

eugenesvk committed:
save on exit

eugenesvk committed:
fmt

eugenesvk committed:
ignore empty cache env vars

eugenesvk committed:
add commands to save/load listed caches

Spitfire1900 commented:
```diff
diff --git a/xonsh/procs/executables.py b/xonsh/procs/executables.py
index ffbe41d4..cc7c4696 100644
--- a/xonsh/procs/executables.py
+++ b/xonsh/procs/executables.py
@@ -324,14 +324,16 @@ class PathCache:  # Singleton
                 lbl += "S" if pn in self.usr_dir_list_session else " "
                 lbl += "M" if pn in self.usr_dir_list_key else " "
                 msg += f"\n {lbl} {p}"
+        check_or_x = "✓" if env.get("XONSH_DIR_CWD_CACHE", False) else "✗"
         msg += (
             "\n\n"
-            + ("✓" if env.get("XONSH_DIR_CWD_CACHE", False) else "✗")
+            + check_or_x
             + " current working dir cache ($XONSH_DIR_CWD_CACHE)\n"
         )
         if env.get("XONSH_DIR_CWD_CACHE", False):
-            msg += (
-                +("✓" if env.get("XONSH_DIR_CWD_CACHE_NON_EXE", False) else "✗")
+            check_or_x = "✓" if env.get("XONSH_DIR_CWD_CACHE_NON_EXE", False) else "✗"
+            msg += (""
+                + check_or_x
                 + " (cwd) cache non-executable ($XONSH_DIR_CWD_CACHE_NON_EXE)"
             )
         if len(self.cwd_too_long):
```

```xonsh
+1@ from xonsh.procs.executables import PathCache; pc = PathCache(None); pc.get_cache_info()
PATH    : ∑  66 dirty
          └  55 clean (unique & existing)
Cached  : ∑   1 of which:              (pc = PathCache(None))
          ├   1 permanently            (pc.usr_dir_list_perma   ← $XONSH_DIR_PERMA_CACHE       )
          ├   0 this session           (pc.usr_dir_list_session ← $XONSH_DIR_SESSION_CACHE     )
          └   0 by dir mtime, list onΔ (pc.usr_dir_list_key     ← $XONSH_DIR_CACHE_TO_LIST     )
Uncached: ∑  54 including:
  C:\Program Files\Git\usr\bin
  C:\Program Files\Git\mingw64\bin
  C:\TMP\xonsh-cacheWinPathPerma\Scripts
  C:\Program Files\Python313\Scripts
  C:\Program Files\Python313
  ...

✓ current working dir cache ($XONSH_DIR_CWD_CACHE)
✓ (cwd) cache non-executable ($XONSH_DIR_CWD_CACHE_NON_EXE)
```

Spitfire1900 commented:
I wonder if non-existent dirs is impacting this?
```xonsh
+1@ _ = [print(f"{pathlib.Path(i).resolve()}, {pathlib.Path(i).is_dir()}") for i in $PATH if pathlib.Path(i).is_dir() is False]

C:\Program Files (x86)\Microsoft SQL Server\150\DTS\Binn, False
C:\Program Files (x86)\Microsoft SQL Server\140\DTS\Binn, False
C:\Program Files (x86)\Microsoft SQL Server\160\DTS\Binn, False
C:\Program Files (x86)\Meld, False
```

eugenesvk commented:
Fixed the wrong `+` concat

Also added experimetnal `$XONSH_DIR_CACHE_LIST_SAVE_ON_EXIT` so you could save the cache to all the listed dirs to a file so you wouldn't list them on every new xonsh session start if no dirs were modified

eugenesvk commented:
Also, try `from xonsh.procs.executables import PathCache; pc = PathCache(None); pc.get_cache_info(2)`, it hints that `pc.get_cache_db('p')` can get you the database for the permacache without having to unpickle yourself, otherwise it's a list of 2 caches


https://github.com/xonsh/xonsh/blob/fff344de7cbf03f0df173a8d1a546ef956eccda1/xonsh/procs/executables.py#L479-L481

eugenesvk commented:
I see there is indeed some bug with the environment being set too early before your `$XONSH_DIR_PERMA_CACHE` is read, for testing maybe move it to the beginning of your rc.xsh?

eugenesvk committed:
update PathCache reset to allow deleting pickled cache files

eugenesvk committed:
make more explicit that usr_dirs come from PathCache

Spitfire1900 commented:
> I see there is indeed some bug with the environment being set too early before your `$XONSH_DIR_PERMA_CACHE` is read

Thanks a lot @eugenesvk;  I'll wait on this before trying again. I should be able to `$XONSH_DIR_PERMA_CACHE = $PATH` in `.xonshrc` and it should handle non-existent folders and Network Drives.

eugenesvk committed:
fmt

eugenesvk committed:
fmt

eugenesvk committed:
don't force-trigger IO path clean updates by calling locate_executable elsewhere

eugenesvk committed:
save env in a cls var

eugenesvk committed:
init PathCache earlier to avoid it being called when is_dirty is reset by a previous call to get_clean_path

eugenesvk committed:
move setting up usr lists of cached dirs to a separate function

to allow using it even after initialization to catch changes in env vars

eugenesvk committed:
fmt

eugenesvk committed:
fix type mismatch

eugenesvk committed:
fix cache reset error

eugenesvk committed:
test: add pathext to not depend on the low number that can appear in CI

Spitfire1900 commented:
Performance still having a lot of issues, this is running a profile over typing the entire alphabet 3 times.

```
@ pstats.Stats('stats.profile').sort_stats(SortKey.CUMULATIVE).print_stats('executables.py:')
Sun Dec 15 16:07:13 2024    stats.profile

         12895477 function calls (12859914 primitive calls) in 76.906 seconds

   Ordered by: cumulative time
   List reduced from 1287 to 20 due to restriction <'executables.py:'>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        5    0.000    0.000   56.548   11.310 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:87(locate_executable)
        5    0.000    0.000   56.548   11.310 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:586(locate_file)
        5    0.001    0.000   56.548   11.310 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:716(locate_file_in_path_env)
        5    0.000    0.000   56.537   11.307 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:470(get_dir_cache_perma)
        5    0.000    0.000   56.537   11.307 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:475(update_cache)
        5    0.012    0.002   56.537   11.307 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:487(_update_paths_cache)
    12270    0.004    0.000   56.351    0.005 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:110(executables_in)
    12270    0.250    0.000   56.347    0.005 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:129(_executables_in_windows)
    65260    0.142    0.000   55.853    0.001 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:46(is_executable_in_windows)
   780200    0.343    0.000    1.183    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:58(<genexpr>)
       10    0.000    0.000    0.004    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:223(get_clean_path)
       10    0.000    0.000    0.004    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:211(get_clean_paths)
        2    0.000    0.000    0.004    0.002 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:704(hash_s_list)
        5    0.000    0.000    0.000    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:613(locate_relative_path)
        5    0.000    0.000    0.000    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:17(get_possible_names)
        5    0.000    0.000    0.000    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:386(__init__)
        5    0.000    0.000    0.000    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:409(set_usr_dir_list)
       10    0.000    0.000    0.000    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:440(cache_file)
        5    0.000    0.000    0.000    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:173(__new__)
        5    0.000    0.000    0.000    0.000 C:\TMP\xonsh-cacheWinPathPerma\src\xonsh\xonsh\procs\executables.py:83(__init__)

```

I only had `$XONSH_DIR_PERMA_CACHE  = [i for i in $PATH]` set during this run.

Spitfire1900 commented:
As a reference with what I'm working with here the following script completes in 42 seconds on my workstation, a recently modern Windows laptop with an on-access virus scanner. On a spinning rust Linux box I have from 2016 it takes ~1 second.

```python
import os
import pathlib

from datetime import datetime

files = 0
iteration = 0
MAX_FILES = 360000
now = datetime.now()
while True:
    print(f'{iteration=}, {files=}')
    for i in pathlib.Path('C:/Program Files/Git/usr/bin/').iterdir():
        stat = os.stat(i)
        files += 1
        if files >= MAX_FILES:
            break
    if files >= MAX_FILES:
        finished = datetime.now()
        break
    else:
        iteration += 1
```

eugenesvk commented:
Yeah, there is already a simple performance test here, which is currently failing on Windows. Think there is some simple failing cache invalidation check that caused the caches to be rebuilt on every keystroke, but haven't investigated it yet

eugenesvk committed:
fix pathext not being cached

eugenesvk commented:
should be better now, no cache vs `$XONSH_DIR_PERMA_CACHE = list($PATH)`
(think it bugs with the EnvPath type as it has no len, but then it's not really meant to  be used like that)

https://github.com/user-attachments/assets/032a8a4d-fbc7-486c-92c9-1aafecdbafd0

Spitfire1900 commented:
That did the trick. On first launch it hung for ~5-7 seconds as it built the cache (presumably) then responded well. Then I re-launched and the hang did not reproduce.

`.xonshrc`
```xonsh
$XONSH_DIR_PERMA_CACHE = list($PATH)
```

At least one of these folders is a network share

```python
@
total_exes = 0
for k, v in pc.get_cache_db('p').items():
    total_exes += len(v)
    print(f'{len(v)=}')
total_folders = len(pc.get_cache_db('p'))
print(f'{total_exes=}')
print(f'{total_folders=}')
...
len(v)=262
len(v)=54
len(v)=18
len(v)=3
len(v)=2
len(v)=35
len(v)=0
len(v)=0
len(v)=3
len(v)=2
len(v)=0
len(v)=0
len(v)=40
len(v)=687
len(v)=14
len(v)=9
len(v)=2
len(v)=9
len(v)=9
len(v)=1
len(v)=1
len(v)=6
len(v)=6
len(v)=6
len(v)=6
len(v)=6
len(v)=0
len(v)=108
len(v)=99
len(v)=15
len(v)=1
len(v)=687
len(v)=14
len(v)=9
len(v)=2
len(v)=9
len(v)=91
len(v)=0
len(v)=2
len(v)=2
len(v)=6
len(v)=16
len(v)=1
len(v)=38
len(v)=8
len(v)=1
len(v)=18
len(v)=58
len(v)=2
len(v)=1
len(v)=7
len(v)=8
len(v)=14
len(v)=1
len(v)=3
total_exes=2402
total_folders=55
```

Spitfire1900 commented:
I'll be using this in lieu of https://github.com/xonsh/xonsh/pull/5677 for awhile.

What is the recommended way to refresh the PERM cache, manually delete the pickle file?

eugenesvk commented:
15 sec is a bit high, but with a slow/large network share seems ok (perma caches do check each file for the existence + executable suffix)?

- You can delete the pickle file directly or
- run this `pc` method with an extra flag to delete both perma and listed caches
```
    def reset(cls, delfiles: bool = False):
        """Clean PathCache to allow creating a new one with a different env
        delfiles: Also delete cached files
        """
```
- cahe also resets if your `pathext` var changes since then the definition of what a command is is different vs what's cached, though that should be optimized on pathext getting "smaller" by removing the extra .exts instead of removing all the cache... so shouldn't be relied on to track changes

But note that perma cache was meant to be... permanent :) Maybe in addition to non-changing dirs like `C:Windows/`you could add **frequently** changing **large** dirs where listing them on changes might be noticeably bad?

But otherwise have you tried moving non-perma dirs to the `XONSH_DIR_CACHE_TO_LIST` cache (with or without `XONSH_DIR_CACHE_LIST_NON_EXE` depending on how much you care about highlight precision)?

It can be pickled on xonsh exit to avoid startup costs with `XONSH_DIR_CACHE_LIST_SAVE_ON_EXIT`, so for rarely changed dirs or smaller dirs that should work?

Spitfire1900 commented:
I'll see about leveraging other env vars but at the moment `$XONSH_DIR_PERMA_CACHE = list($PATH)` is the only thing I'm going to have set, then just delete the pickle file once a quarter or so. I may explore the other options once it bites me.

In practice I think that's going to be used most often.

eugenesvk committed:
skip checking for perma-cache updates on every keystroke (respect is_dirty)

eugenesvk committed:
on shrinking pathext don't invalidate the whole cache, but instead remove the extra elements

add shrink_dir_cache_perma to handle it

eugenesvk committed:
add a bit of color to the cache info

eugenesvk committed:
add a help_me_choose method for a rouch speed comparison of caching methods

eugenesvk committed:
fmt

eugenesvk committed:
add ColorXonsh to allow abbreviated use of colors

eugenesvk committed:
doc: update env var description

eugenesvk committed:
add current cache status indicator

eugenesvk committed:
avoid dupe file exists

eugenesvk committed:
update comments in help_me_choose

eugenesvk committed:
add mtime calc and fix path.ext time calculations in PathCache's help_me_choose

eugenesvk committed:
add env var for XONSH_DIR_CACHE_TO_ALIST

eugenesvk committed:
update comment

eugenesvk committed:
fmt

eugenesvk committed:
fmt

eugenesvk committed:
add non-executable listed dirs to a separate cache list

eugenesvk committed:
fmt

eugenesvk committed:
update comments

eugenesvk committed:
types

eugenesvk committed:
fix arg error

Spitfire1900 commented:
> > That did the trick. On first launch it hung for ~15 seconds
> 
> wonder, is there an easy way to do that work when the user isn't typing anything at the prompt instead of blocking the UI for so long?

Delegate auto-completion related tasks off of the main thread.

eugenesvk committed:
add suggestions to help_me_choose

eugenesvk committed:
remove ref to unimplemented XONSH_DIR_SESSION_CACHE_SHARE

eugenesvk committed:
fmt

eugenesvk committed:
fix wrong types

eugenesvk commented:
Btw, I've added a `help_me_choose` method to offer suggestions on which cache dirs from `$PATH` could be added.

`from xonsh.procs.executables import PathCache; pc = PathCache; pc.help_me_choose()`

eugenesvk committed:
fix mypy

eugenesvk committed:
add Nix perma suggestions

eugenesvk committed:
don't quit on Perma suggestions, check if other caches make sense

eugenesvk committed:
update suggestions fmt

eugenesvk committed:
add an absolute timing check for suggestions

even without pathext if listing is fast (but slower than pathext) it's ok to list on new prompt but get prefix highlighting

eugenesvk committed:
fmt

eugenesvk committed:
unblock UI: move perma cache updates off the main thread

eugenesvk committed:
typo

eugenesvk committed:
comment

eugenesvk committed:
add missing vars to get_cache_info, fmt

eugenesvk committed:
update description

eugenesvk committed:
fmt

eugenesvk committed:
rename var to match cache types

dir_key_cache → dir_cache_listed
dir_cache → dir_cache_sess

eugenesvk commented:
and moved cache updates off the main thread, so hopefully no more huge delays on startup

eugenesvk committed:
remove needless indirection

Spitfire1900 commented:
> and moved cache updates off the main thread, so hopefully no more huge delays on startup

This change worked wonders :) 

On the aside, it looks like $PATH entries on a network drive causes `pc.help_me_choose()` to hang when interrogating that entry
```xsh
14·ext   list_exe   list_all  # files  # execs   mtime     CacheIn    Cached?
...
 .   6    .   2    .   1           3 =     3  2024-01-18        A   ✓ P      C:\Program Files\Python312\Scripts
 .   6    .   6    .   1           8       2  2024-01-18      L     ✓ P      C:\Program Files\Python312
### next entry is a Network drive, hangs here.
```

eugenesvk commented:
It tries to check every path 10 times using various methods unless the 1st list results in many items, but not sure this has any benefit tbh, so might change all to a single list for all dirs. Though in theory this should just take a lot of time, not hang?

You could check by calling `def help_me_choose(cls, iters: int = 10):` with iters set to `1`? This should still do a few per-every-file calls to assess various methods, but hopefully should complete in a more reasonable time

Spitfire1900 commented:
It ended up just taking a long time

```xsh
🐚 N:\NetShare\CmdUtils\bin     24-12-21 12:44:34-0
@ from datetime import datetime; print(datetime.now()); from xonsh.procs.executables import PathCache; pc = PathCache; pc.help_me_choose(); print(datetime.now())
2024-12-21 12:45:23.248089
...
14·ext   list_exe   list_all  # files  # execs   mtime     CacheIn    Cached?
 .   6    . 256    .   3         397     264  2023-08-22    S?      ✓ P      C:\Program Files\Git\usr\bin
 .   6    . 108    .   2         161      56  2023-08-22    S?      ✓ P      C:\Program Files\Git\mingw64\bin
 .   6    .   2    .   1           3 =     3  2024-11-21        A   ✓ P      C:\Program Files\Python313\Scripts
 .   6    .   6    .   1           8       2  2024-11-21      L     ✓ P      C:\Program Files\Python313
 .  19   12.0333    .   6          55      40  2020-06-30    S?      ✓ P     N:\NetShare\CmdUtils\bin # Netshare on PATH enforced by GPO
...
 .   6    .  12    .   1          17       2  2023-08-22      L?    ✓ P      C:\Program Files (x86)\Notepad++
 .  19   11.9524    .   4          55      40  2020-06-30    S?      ✓ P      .
2024-12-21 12:49:53.606617
```

eugenesvk commented:
Do you know how to detect `N:\NetShare\CmdUtils\bin` as a network drive?

Spitfire1900 commented:
> and moved cache updates off the main thread, so hopefully no more huge delays on startup

This brings up a higher level discussion, should the entire completers system in Xonsh be moved off of the main thread and is that feasible?

eugenesvk commented:
That's the tricky part I mentioned ealier. There was an attempt for pyghooks syntax highlighting https://github.com/xonsh/xonsh/pull/4328/files following https://github.com/xonsh/xonsh/issues/2578, but that stumbled since threads aren't interruptable (though there was a sensible suggestion to use just 1 and if it's unavailable, do nothing?)

But otherwise haven't looked into it, not sure to what extent it's feasible and whether there are any bigger issues

would be nice if you never got ui-blocked on typing...

Spitfire1900 commented:
> Do you know how to detect `N:\NetShare\CmdUtils\bin` as a network drive?

I used https://discourse.mcneel.com/t/system-io-is-missing-driveinfo-in-python2-rhino8/175458 as a reference.

```python
# pip install pythonnet
import clr
clr.AddReference('System.Management')  # Add reference for WMI access
from System.Management import ManagementObjectSearcher

def get_drive_type(type_number):
    types = {
        0: "Unknown",
        1: "No Root Directory",
        2: "Removable Disk",
        3: "Local Disk",
        4: "Network Drive",
        5: "Compact Disc",
        6: "RAM Disk"
    }
    return types.get(int(type_number), "Unknown")


searcher = ManagementObjectSearcher('SELECT * FROM Win32_LogicalDisk')

for disk in searcher.Get():
    print(f'{disk['DeviceID']=}, type: {get_drive_type(disk['DriveType'])=}')

```
Outputs:
```
disk['DeviceID']='C:', type: get_drive_type(disk['DriveType'])='Local Disk'
disk['DeviceID']='N:', type: get_drive_type(disk['DriveType'])='Network Drive'
```

eugenesvk committed:
don't bench slow dirs multiple times similar to large dirs

eugenesvk commented:
I've meanwhile excluded dirs that take > 5 secs from doing the iterations multiple times similar to > 1000 items dirs

Spitfire1900 commented:
A network drive test is only useful to present to the user that you're working on benching the entry and not stuck. Though I think it's generally better to just tell the user that some entries may take a long time to benchmark if they're network drives.

Spitfire1900 commented:
So I really like this feature and this branch is my primary installation on my workstation, but on my personal machine I don't have a reason to adopt these features cause I don't have a performance ruining on-access virus scanner installed.

It was mentioned much earlier that there were serious concerns preventing this from being pulled into main, perhaps this is better left as a completers xontrib than in the main project?

eugenesvk committed:
add a slow warning

eugenesvk commented:
What's you video of a held key like on a personal machine with/without cache? For me xonsh has always lagged, those hundreds of IO calls do accumulate on fast typing/deletes if it's all sub-second.

But even outside of perf the new prefixed highlight is great for catching typos early, so I'm using this on non-Win as well.

This mods the core locate executable/file functions, so no idea how to do this in a xontrib

eugenesvk commented:
> Linux box and I could not detect lag.

that's because only Windows does an order of magnitude more checks due to pathext... (note sure, but maybe also Wins have more dirs in paths?)

Yes, the underline, try to type `anit` and see it disapper

Spitfire1900 commented:
With the change to make caching happen off of the main thread I've changed my configuration to use the session cache instead of the permanent cache, `$XONSH_DIR_SESSION_CACHE = list($PATH)`.

Spitfire1900 commented:
Oh man, GPO is enforcing both `'C:\\windows\\system32'` and `'C:\\WINDOWS\\system32'` on my PATH, resulting in two cache entries for it. 

While generally Windows is not case sensitive, the fact that this is toggle-able at the directory level means that the caching code is handling this how it should, despite the double-entry.

https://learn.microsoft.com/en-us/windows/wsl/case-sensitivity

Spitfire1900 commented:
> that's because only Windows does an order of magnitude more checks due to pathext... (note sure, but maybe also Wins have more dirs in paths?)

I have 4604 files in System32, of which 669 are considered executable, I wanted to say that was the reason, alas Linux has many more.

Cumulatively though on my Windows install amongst 56 folders I have 12,403 total executables

Linux: 
```
1·ext   list_exe   list_all  # files  # execs   mtime     CacheIn    Cached?
 .        .   2    .              15 =    15  2024-12-19        A    ✗       /home/kyle/.cargo/bin
 .        .   5    .   2          21 =    21  2024-11-07        A    ✗       /home/kyle/.local/bin
 .        .        .               0 =     0  2015-10-21        A    ✗       /usr/local/sbin
 .        .        .               0 =     0  2024-03-02        A    ✗       /usr/local/bin
 .        .  73    .   5         650 =   650  2024-12-17  P?    A    ✗       /usr/sbin
 .        . 341    .  30        3058 =  3058  2024-12-23  P?    A    ✗       /usr/bin
 .        .  70    .   5         650 =   650  2024-12-17  P?    A    ✗       /sbin
 .        . 318    .  27        3058 =  3058  2024-12-23  P?    A    ✗       /bin
 .        .   1    .              14 =    14  2024-11-05        A    ✗       /usr/games
 .        .        .               0 =     0  2015-10-21        A    ✗       /usr/local/games
 .        .   1    .               8       7  2024-12-23             ✗       /snap/bin
 .        .        .               1 =     1  2023-09-21        A    ✗       //home/kyle/.pyenv/bin
```

eugenesvk commented:
The number of files doesn't matter for the original lag, only 
- pathext * dirs in path

because that's what's being checked, each possible variant you type, the dirs aren't listed

And you have 1 pathext on nix

For caching it matters, but then only for the initial/update on change (and the huge Windows system 32 is recommended to be in permacache, so it doesn't even update

eugenesvk commented:
I thought there was some normalization/deduping going on as part of path cleanup routines, but I guess not

In general,  think case sensitivity should be enabled separately as it's too rare to have on by default

Spitfire1900 commented:
I have a new xontrib created in the simliar vein as this, [xontrib-pygitstatus](https://pypi.org/project/xontrib-pygitstatus/) which eliminates calls to git to populate the PROMPT, using PyGit2 instead.

Spitfire1900 commented:
This is in a pretty good state at this point and is my daily runner on Windows, I've not needed on Linux.

@anki-code may merging this in be re-visited?

eugenesvk committed:
skip save_cache_listed instead of aborting if no pathcache was initialized

Spitfire1900 commented:
Originally the primary root problem for me was a very aggressive, corporate anti-malware program that  would block on all file opens until the virus scanner finished scanning the file making any Python program very slow to start, cause pip builds to fail because it would time out on temporary workspace deletion (the scanner wanted to scan the file before you could delete it 🤦‍♂️)and cause XONSH to be slow to type unless I disabled executable highlighting.

anki-code commented:
Outdated

eugenesvk commented:
How is it outdated if the issue it solves remains in the latest version?

Spitfire1900 commented:
@eugenesvk I'd argue having so many executables the startup impact is a concern or having valid executables added after startup that could benefit from short-lived caches that are not session long are out-of-scope.

eugenesvk commented:
So what's your argument given that any default Windows installation has a lot of executables?

Or why wouldn't you want a newly created user executable from showing up in your shell right away without a session restart?

Spitfire1900 commented:
I use xonsh on Windows as my primary work environment, xonsh is way better at handling this case on the latest feature version than last year.
